math/src/main/java/smile/math/special/Beta.java

```diff
@@ -54,27 +54,27 @@ public static double beta(double x, double y) {
      * Continued Fraction approximation (see Numerical recipies for details)
      */
     public static double regularizedIncompleteBetaFunction(double alpha, double beta, double x) {
-        if (x < 0.0 || x > 1.0) {
+        double EPS = 1e-8;
+        if (x + EPS < 0.0 || x - EPS > 1.0) {
```

Comment:
This is incorrect. x should be in (0, 1). This change will allow invalid x mathematically.

Reply:
It only allows values that are 1e-8 less than zero or 1e-8 greater than 1, to account for floating point arithmetic errors. It's a standard practice to account for such errors when comparing floating point numbers: https://stackoverflow.com/questions/1088216/whats-wrong-with-using-to-compare-floats-in-java

The range of possible values for x is inclusive of 0 and 1, right? The check here is for < and >, not <= and >=; plus the x == 0.0 and x==1.0 conditions are handled below...

Now that I've thought about it more, this is a more straightforward implementation:
```
        if (Math.abs(x - 0.0) < EPS) {
            return 0.0;
        }

        if (Math.abs(x - 1.0) < EPS) {
            return 1.0;
        }

        if (x < 0.0 || x > 1.0) {
            throw new IllegalArgumentException("Invalid x: " + x);
        }
```
If after reviewing my reasoning, you'd like to make the change and you prefer the latter method, I can push another commit.

Reply:
Yes, it is inclusive and B(0,a,b)=0 and B(1,a,b)=1 always. The latter is much better. Thanks

Reply:
I will add this to v2. no need more PR. Thanks